tools/generator/src/Generator+DistinguishTargets.swift

```diff
@@ -0,0 +1,257 @@
+    /// name, uniquely distinguishes it from targets with the same non-
+    /// distinguished name.
+    ///
+    /// - Precondition: All targets have been added with `add(target:)` before
+    ///   this is called.
+    ///
+    /// - Parameters:
+    ///   - target: The `Target` to generate a distinguisher for.
+    ///   - forceDistinguisher: If `true`, the operating system name will be
+    ///     part of the string returned.
+    ///
+    /// - Returns: `nil` if no distinguisher is needed.
+    func distinguisher(target: Target, forceDistinguisher: Bool) -> String? {
+        let platform = target.platform
+
+        if distinguisherKeys[target.distinguisherKey]! > 1 {
+            // The target name would be ambiguous, even with our distinguisher
+            // components. We will show a shorted configuration hash instead,
+            // which will be unique.
+            var components: [String] = []
+
+            if forceDistinguisher {
+                components.append(platform.os)
+            }
+
+            components.append(Target.prettyConfiguration(target.configuration))
+
+            return components.joined(separator: ", ")
+        } else {
+            var components: [String] = []
+
+            if archs.count > 1 {
+                components.append(platform.arch)
+            }
+            if forceDistinguisher || minimumVersions.count > 1 {
+                components.append(platform.os)
+            }
+            if minimumVersions.count > 1 {
+                components.append(platform.minimumOsVersion)
+            }
+            if environments.count > 1 {
+                components.append(platform.environment ?? "Device")
+            }
+
+            return components.isEmpty ? nil : components.joined(separator: " ")
+        }
+    }
+}
+
+extension Target {
+    /// Memoized configuration hashes.
+    private static var configurationHashes: [String: String] = [:]
+
+    /// Returns a user-facing string for a given configuration.
+    static func prettyConfiguration(_ configuration: String) -> String {
+        if let hash = configurationHashes[configuration] {
+            return hash
+        }
+
+        let hash = String(configuration.md5Hash().prefix(5))
+        configurationHashes[configuration] = hash
+
+        return hash
+    }
+}
+
+private extension Target {
+    /// A key that corresponds to the most-distinguished string that
+    /// `OperatingSystemComponents.distinguisher()` can return for this
+    /// `Target`.
+    var distinguisherKey: String {
+        // This doesn't need `platform.os`, as `OperatingSystemComponents` is
+        // already segregated by operating system name.
+        return [
+            platform.arch,
+            platform.minimumOsVersion,
+            platform.environment ?? "Device"
+        ].joined(separator: "-")
+    }
+}
+
+private extension String {
+    /// Computes an MD5 hash string for this `String`.
```

Comment:
I recently read that md5 is slower than sha1 on modern cpus, so I'll adjust this.